Skip Checks API neutral state

yamachu opened the pull request:
We use CircleCI for CI.
Recently it will returns a `neutral` state before state is determined by the CI.
In the current code, except for `success`, it is designed to be considered a failure.
https://github.com/voyagegroup/popuko/blob/71080116e305bb469dd35732e11a0199fdaa4ecd/epic/check_autobranch.go#L186
I thought it would be better to skip the decision for `neutral`, which cannot be determined as success or failure, so I added that process.

<img width="458" alt="スクリーンショット 2021-11-15 18 51 04" src="https://user-images.githubusercontent.com/1955233/141760313-4199c2b6-bf41-457f-88cb-749c6209ab3a.png">

Webhook payload
```
{
  "action": "completed",
  "check_suite": {
    "id":,
    "node_id": "",
    "head_branch": "master",
    "head_sha": "",
    "status": "completed", <- completed???? test is running...
    "conclusion": "neutral", <- 
    "url": "",
    "before": "",
    "after": "",
    "pull_requests": [

    ],
   ...
}
```

yamachu committed:
The neutral status cannot be determined as success or failure, so the process is skipped

For example, CircleCI may often send it

tetsuharuohzeki commented on epic/check_autobranch.go:
```diff
@@ -183,6 +183,11 @@ func mergeSucceedItem(
 		return true
 	}
 
+	if info.Status == "neutral" {
```
I just knew [`conclusion`](https://docs.github.com/en/rest/reference/checks#create-a-check-run) field now....

I think we should define `StateChangeInfo.Conclusion` and check it at here.

By #261, we fill `StateChangeInfo.Status` in two way:

- https://github.com/voyagegroup/popuko/blob/c1901c2b115bbd84e213e83d029df0e3c78b392e/epic/check_autobranch.go#L27
- https://github.com/voyagegroup/popuko/blob/c1901c2b115bbd84e213e83d029df0e3c78b392e/epic/check_autobranch.go#L41

But `github.StatusEvent.Status` and `github.StatusEvent.Conclusion` are different field. It's time to stop to mix them into single field. Let's handle them separately.

yamachu replied:
> `github.StatusEvent.Status` and `github.StatusEvent.Conclusion` are different field. It's time to stop to mix them into single field.

Yes, from code design point of view, that's a good idea.
However, some refactoring is considered necessary.

BTY, I'm sorry to postpone the issue, but how about the following changes.

```diff
- NotHandle                 bool
+ InProgress                 bool
```

```diff
- NotHandle:                 *ev.CheckSuite.Status != "completed",
+ InProgress:                 *ev.CheckSuite.Status != "completed" || *ev.CheckSuite.Conclusion == "neutral"
```

tetsuharuohzeki replied:
@yamachu 

Even though, I suspect it's better to add `StateChangeInfo.InProgress()` method instead of your proposed approach.
It's nice abstraction to push something into the appropriate method. I think that changing struct field as your proposed is not robustness for the future. 

Then we would need to change to hold values of `github.StatusEvent.Status` and `github.StatusEvent.Conclusion` in `StateChangeInfo`. This requires a similar refactoring as I told the above.

yamachu committed:
Add method to determine CI is in progress or not

tetsuharuohzeki commented on epic/check_autobranch.go:
```diff
@@ -20,6 +20,7 @@ type StateChangeInfo struct {
 	ID                        int64
 	SHA                       string
 	IsRelatedToAutoBranchBody func(string) bool
+	IsInProgress              func() bool
```
@yamachu I imagined the following code instead of passing a function to `.IsInProgress` field.

```diff
type StateChangeInfo struct {
+    Conclusion string
}

+func IsInProgress(s *StateChangeInfo) bool {
+  ....
+}
```

What do you think about this? Are there any problem to implement?

yamachu replied:
@tetsuharuohzeki In the code you were suggesting, did you imagine to set the Conclusion field to `ev.State` when received a StatusAPI event?

tetsuharuohzeki replied:
Yes.

Ah, but `github.StatusEvent` does not have `Conclusion` field. I think we can use `*string` for `StateChangeInfo.Conclusion`

tetsuharuohzeki replied:
@yamachu 

> Ah, but github.StatusEvent does not have Conclusion field. I think we can use *string for StateChangeInfo.Conclusion

I wrote that but I also seem your proposed code would also be nice approach which uses func.
You can choose the final implementation. How do you think?

yamachu replied:
It's a different types, but I think I can express it in a function like TypeScript's Union type, so I'll add a `Conclusion 
 *string` parameter.

In the following comments, would such an implementation be appropriate?
https://github.com/voyagegroup/popuko/pull/278/files#r754050176

```go
func isStatusDetermined(info StateChangeInfo) bool {
	// Status API
	if info.Conclusion == nil {
		return true // or info.Status != "pending"
	}

	// Checks API
	return *info.Conclusion != "neutral"
}

func isStatusSuccess(info StateChangeInfo) bool {
	// Status API
	if info.Conclusion == nil {
		return info.Status == "success"
	}

	// Checks API
	return *info.Conclusion == "success"
}
```

tetsuharuohzeki replied:
@yamachu 

Umm, I agree that we would need to define a status by something.

However, from your proposed psuedo code, I think _the current design_ is wrong to handle these status.
Your proposed one mixes handling _Status API_ and _Checks API_ in the same code path. This makes a thing more complex.

I think we would require these refactoring to introduce this PR change cleanly:

1. Introduce interface type which abstracts `StateChangeInfo`, and separate a code path into for  _Status API_ and _Checks API_.
2. Do something which we would like to do in this pull request.

yamachu replied:
私は英語ネイティブではないため、コメントに対してコストがかかってしまっています。
そのためコメントを日本語で記載いたします。

--- 

異なる概念であるStatus APIのEventとChecks APIのEventを同一の構造体で表現していることで複雑になっていることは同意いたします。
例えば `info.Status` で表現されている内容が、ChecksAPIの場合はConclusionの値であるといった例です。

そのためコードパスを分離することは確かに良いように思われます。
Event依存の構造体に非依存の構造体を入れ込み、Event依存のコードパスを分離することでまずは表現可能だと考え、以下のコミットで表現してみました。
https://github.com/voyagegroup/popuko/pull/278/commits/1769fdedaac763958f7158d925cd44e330c1d997

非依存の構造体に例えば `IsRelatedToAutoBranchBody` のように、関数を返す形にすることで

```go
log.Println("info: the status event is related to auto branch.")

info.mergeSucceedItem(ctx, client, info.Owner, info.Name, repoInfo, q)

q.RemoveActive()
```

の様にエントリポイントである `checkAutoBranch` は共通化出来そうではありますが、この構造体に持たせるのは違和感がありました。
そのため横に別名で生やすことで、コードの分離を実現しています。

tetsuharuohzeki commented on epic/check_autobranch.go:
```diff
@@ -183,6 +198,11 @@ func mergeSucceedItem(
 	}
 
+	if info.IsInProgress() {
+		log.Printf("info: Check is in progres\n")
```
nit: `progress`?

tetsuharuohzeki commented on epic/check_autobranch.go:
```diff
@@ -130,6 +133,18 @@ func isRelatedToAutoBranchBodyWithCheckSuiteEvent(ev *github.CheckSuiteEvent) fu
+
+func inProgressWithCheckSuiteEvent(ev *github.CheckSuiteEvent) func() bool {
+	return func() bool {
+		return *ev.CheckSuite.Status != "completed" || *ev.CheckSuite.Conclusion == "neutral"
```
@yamachu 

[As your comment](https://github.com/voyagegroup/popuko/pull/278#issue-1053455717), I feel `*ev.CheckSuite.Conclusion == "neutral"` does not mean _in progress_ state directly.

But this code will check whether `*ev.CheckSuite.Conclusion` is `neutral` after if `*ev.CheckSuite.Status != "completed"` equals to that `*ev.CheckSuite.Status` indicates some stopped states. 

I concern that this function will return `true` in almost case. Is this right?

yamachu replied:
> I concern that this function will return true in almost case. Is this right?

Yes, it enforces the order of use, which doesn't seem to be a good function...
Rather, what should be expressed is not whether it is in progress or not, but whether the state is determined or not.

yamachu committed:
Separate Status and CheckSuite Event struct

yamachu committed:
Remove unused method

No longer needed as the same processing will be done within each Event-dependent process

yamachu committed:
Duplicate checkAutoBranch for each Event

yamachu committed:
The neutral status cannot be determined as success or failure, so the process is skipped

tetsuharuohzeki commented:
@yamachu Sorry, I might not be able to reply soon....

tetsuharuohzeki commented on epic/check_autobranch.go:
```diff
@@ -11,8 +11,8 @@ import (
 	"github.com/voyagegroup/popuko/setting"
 )
 
+// FIXME: Name
```
@yamachu Would you fix this comment before merging this?

tetsuharuohzeki reviewed (commented):
@yamachu 

Thank you for your patience. I seem your new changesets are nice. 

Probably, you may want to squash these commits or reorganize commits. But I feel other points are no problems and ready to merge nearly.

- https://github.com/voyagegroup/popuko/pull/278/commits/c6d909aef7dc407cbb3246d85061a0c88a6bc39e 
- https://github.com/voyagegroup/popuko/pull/278/commits/9e5bafff91e993e83aaa7c782226d51f00f31fd5

tetsuharuohzeki commented:
@yamachu Do you still face this problem?
